Classes/common/Encryption/Keychain/CDTEncryptionKeychainManager.m

```diff
@@ -0,0 +1,193 @@
+
+        if (![self.storage saveEncryptionKeyData:keychainData]) {
+            dpk = nil;
+        }
+    }
+
+    return dpk;
+}
+
+- (BOOL)encryptionKeyDataAlreadyGenerated { return [self.storage encryptionKeyDataExists]; }
+
+- (BOOL)clearEncryptionKeyData { return [self.storage clearEncryptionKeyData]; }
+
+#pragma mark - CDTEncryptionKeychainManager+Internal methods
+- (BOOL)isValidEncryptionKeyData:(CDTEncryptionKeychainData *)data
+{
+    // Ensure the num derivations saved, matches what we have
+    if (data.iterations != CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT,
+                   @"Number of stored iterations does NOT match the constant value %li",
+                   (long)CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS);
+
+        return NO;
+    }
+
+    // Ensure IV has the correct length
+    if (data.iv.length != CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE) {
+        CDTLogWarn(CDTDATASTORE_LOG_CONTEXT, @"IV does not have the expected size: %i bytes",
+                   CDTENCRYPTION_KEYCHAIN_AES_IV_SIZE);
+
+        return NO;
+    }
+
+    return YES;
+}
+
+- (NSData *)generateDpk
+{
+    NSData *dpk = [CDTEncryptionKeychainUtils
+        generateRandomBytesInBufferWithLength:CDTENCRYPTION_KEYCHAIN_ENCRYPTIONKEY_SIZE];
+
+    return dpk;
+}
+
+- (CDTEncryptionKeychainData *)keychainDataToStoreDpk:(NSData *)dpk
+                                encryptedWithPassword:(NSString *)password
+{
+    NSData *salt = [self generatePBKDF2Salt];
+
+    NSData *key =
+        [self generateAESKeyUsingPBKDF2ToDerivePassword:password
+                                               withSalt:salt
+                                             iterations:CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS
+                                                 length:CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE];
+
+    NSData *iv = [self generateAESIv];
+
+    NSData *encryptedDpk = [self encryptDpk:dpk usingAESWithKey:key iv:iv];
+
+    CDTEncryptionKeychainData *keychainData =
+        [CDTEncryptionKeychainData dataWithEncryptedDPK:encryptedDpk
+                                                   salt:salt
+                                                     iv:iv
+                                             iterations:CDTENCRYPTION_KEYCHAIN_PBKDF2_ITERATIONS
+                                                version:CDTENCRYPTION_KEYCHAIN_VERSION];
+
+    return keychainData;
+}
+
+- (NSData *)generatePBKDF2Salt
+{
+    NSData *salt = [CDTEncryptionKeychainUtils
+        generateRandomBytesInBufferWithLength:CDTENCRYPTION_KEYCHAIN_PBKDF2_SALT_SIZE];
+
+    return salt;
+}
+
+- (NSData *)generateAESKeyUsingPBKDF2ToDerivePassword:(NSString *)password
+                                             withSalt:(NSData *)salt
+                                           iterations:(NSInteger)iterations
+                                               length:(NSUInteger)length
```

Comment:
Are you talking about `CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE`? In this contrary, on the contrary to the size of IV, there are 3 valid values: 16, 24 & 32.